File: app/src/main/java/com/tunes/player/playback/PlaybackManager.java
package com.tunes.player.playback;

import android.content.Context;
import android.media.MediaMetadata;
import android.media.session.PlaybackState;
import android.os.Bundle;
import android.os.SystemClock;

import com.tunes.player.model.MusicModel;
import com.tunes.player.singleton.TrackManager;

public class PlaybackManager implements Playback.Callback {

    // Constants used by MediaNotificationManager and other classes to read metadata
    public static final String METADATA_TITLE_KEY    = MediaMetadata.METADATA_KEY_TITLE;
    public static final String METADATA_ARTIST_KEY   = MediaMetadata.METADATA_KEY_ARTIST;
    public static final String METADATA_ALBUM_KEY    = MediaMetadata.METADATA_KEY_ALBUM;
    public static final String METADATA_ALBUM_ART    = MediaMetadata.METADATA_KEY_ALBUM_ART;
    public static final String METADATA_DURATION_KEY = MediaMetadata.METADATA_KEY_DURATION;

    // Direction constants used by TrackManager.canSkipTrack()
    public static final short ACTION_PLAY_NEXT = 1;
    public static final short ACTION_PLAY_PREV = -1;

    public interface PlaybackServiceCallback {
        void onPlaybackStart();
        void onPlaybackStopped();
        void onStartNotification();
        void onStopNotification();
        void onPlaybackStateChanged(PlaybackState newState);
        void onMetaDataChanged(MediaMetadata newMetaData);
    }

    private final Context mContext;
    private final Playback mPlayback;
    private final TrackManager mTrackManager;
    private final PlaybackServiceCallback mServiceCallback;
    private int mLastKnownMediaId = -1;

    public PlaybackManager(Context context, Playback playback,
                           TrackManager trackManager,
                           PlaybackServiceCallback serviceCallback) {
        this.mContext = context;
        this.mPlayback = playback;
        this.mTrackManager = trackManager;
        this.mServiceCallback = serviceCallback;
        mPlayback.setCallback(this);
    }

    // ── MediaSession.Callback ──────────────────────────────────────────────

    public android.media.session.MediaSession.Callback getSessionCallbacks() {
        return new android.media.session.MediaSession.Callback() {

            @Override
            public void onPlay() {
                handlePlay();
            }

            @Override
            public void onPause() {
                handlePause();
            }

            @Override
            public void onStop() {
                handleStop();
            }

            @Override
            public void onSkipToNext() {
                handleSkipToNext();
            }

            @Override
            public void onSkipToPrevious() {
                handleSkipToPrev();
            }

            @Override
            public void onSeekTo(long pos) {
                mPlayback.onSeekTo(pos);
            }

            @Override
            public void onPlayFromMediaId(String mediaId, Bundle extras) {
                handlePlay();
            }
        };
    }

    // ── Private handlers ───────────────────────────────────────────────────

    private void handlePlay() {
        MusicModel item = mTrackManager.getActiveQueueItem();
        if (item == null) return;

        boolean mediaChanged = (item.getId() != mLastKnownMediaId);
        mLastKnownMediaId = item.getId();

        mPlayback.onPlay(item, mediaChanged);
        mServiceCallback.onPlaybackStart();
        mServiceCallback.onStartNotification();

        if (mediaChanged) {
            publishMetadata(item);
        }
    }

    private void handlePause() {
        mPlayback.onPause();
        mServiceCallback.onStopNotification();
    }

    private void handleStop() {
        mPlayback.onStop(true);
        mServiceCallback.onPlaybackStopped();
        mServiceCallback.onStopNotification();
    }

    private void handleSkipToNext() {
        if (mTrackManager.canSkipTrack(ACTION_PLAY_NEXT)) {
            handlePlay();
        } else {
            handleStop();
        }
    }

    private void handleSkipToPrev() {
        if (mPlayback.getCurrentStreamingPosition() > 3000) {
            // If more than 3s in, restart current track
            mPlayback.onSeekTo(0);
        } else if (mTrackManager.canSkipTrack(ACTION_PLAY_PREV)) {
            handlePlay();
        }
    }

    private void publishMetadata(MusicModel md) {
        MediaMetadata.Builder builder = new MediaMetadata.Builder()
                .putString(METADATA_TITLE_KEY,  md.getSongName())
                .putString(METADATA_ARTIST_KEY, md.getArtist())
                .putString(METADATA_ALBUM_KEY,  md.getAlbum())
                .putLong(METADATA_DURATION_KEY, md.getDuration());

        mServiceCallback.onMetaDataChanged(builder.build());
    }

    // ── Playback.Callback ──────────────────────────────────────────────────

    @Override
    public void onPlaybackCompletion() {
        handleSkipToNext();
    }

    @Override
    public void onPlaybackStateChanged(int state) {
        PlaybackState.Builder builder = new PlaybackState.Builder();
        long actions = PlaybackState.ACTION_PLAY
                | PlaybackState.ACTION_PAUSE
                | PlaybackState.ACTION_PLAY_PAUSE
                | PlaybackState.ACTION_SKIP_TO_NEXT
                | PlaybackState.ACTION_SKIP_TO_PREVIOUS
                | PlaybackState.ACTION_STOP
                | PlaybackState.ACTION_SEEK_TO;

        builder.setActions(actions)
               .setState(state, mPlayback.getCurrentStreamingPosition(),
                         1.0f, SystemClock.elapsedRealtime());

        mServiceCallback.onPlaybackStateChanged(builder.build());
    }

    @Override
    public void onFocusChanged(boolean resumePlayback) {
        if (resumePlayback) {
            handlePlay();
        } else {
            handlePause();
        }
    }
}

File: app/src/main/java/com/jegly/frequency/playback/PlaybackManager.kt
package com.jegly.frequency.playback

import android.content.Context
import android.media.MediaMetadata
import android.media.session.MediaSession
import android.media.session.PlaybackState
import android.os.Bundle
import android.os.SystemClock
import com.jegly.frequency.utils.AppSettings
import com.jegly.frequency.model.MusicModel
import com.jegly.frequency.singleton.TrackManager

class PlaybackManager(
    private val context: Context,
    private val playback: Playback,
    private val trackManager: TrackManager,
    private val serviceCallback: PlaybackServiceCallback
) : Playback.Callback {

    companion object {
        const val ACTION_SET_SPEED = "com.jegly.frequency.ACTION_SET_SPEED"
        const val METADATA_TITLE_KEY = MediaMetadata.METADATA_KEY_TITLE
        const val METADATA_ARTIST_KEY = MediaMetadata.METADATA_KEY_ARTIST
        const val METADATA_ALBUM_KEY = MediaMetadata.METADATA_KEY_ALBUM
        const val METADATA_ALBUM_ART = MediaMetadata.METADATA_KEY_ALBUM_ART
        const val METADATA_DURATION_KEY = MediaMetadata.METADATA_KEY_DURATION

        const val ACTION_PLAY_NEXT: Short = 1
        const val ACTION_PLAY_PREV: Short = 2
    }

    interface PlaybackServiceCallback {
        fun onPlaybackStart()
        fun onPlaybackStopped()
        fun onStartNotification()
        fun onStopNotification()
        fun onPlaybackStateChanged(newState: PlaybackState)
        fun onMetaDataChanged(newMetaData: MediaMetadata)
    }

    private var lastKnownPath = ""

    init {
        playback.setCallback(this)
    }

    val sessionCallbacks: MediaSession.Callback
        get() = object : MediaSession.Callback() {
            override fun onPlay() {
                handlePlay()
            }

            override fun onPause() {
                handlePause()
            }

            override fun onStop() {
                handleStop()
            }

            override fun onSkipToNext() {
                handleSkipToNext()
            }

            override fun onSkipToPrevious() {
                handleSkipToPrev()
            }

            override fun onSeekTo(pos: Long) {
                playback.onSeekTo(pos)
            }

            override fun onCustomAction(action: String, extras: Bundle?) {
                if (action == ACTION_SET_SPEED) {
                    val speed = extras?.getFloat("speed", 1.0f) ?: 1.0f
                    playback.setPlaybackSpeed(speed)
                }
            }

            override fun onPlayFromMediaId(mediaId: String?, extras: Bundle?) {
                handlePlay()
            }
        }

    private fun handlePlay() {
        val item = trackManager.getActiveQueueItem() ?: return
        val mediaChanged = item.songPath != lastKnownPath
        lastKnownPath = item.songPath

        playback.onPlay(item, mediaChanged)
        serviceCallback.onPlaybackStart()
        serviceCallback.onStartNotification()

        if (mediaChanged) {
            publishMetadata(item)
        }
    }

    private fun handlePause() {
        playback.onPause()
    }

    private fun handleStop() {
        playback.onStop(true)
        serviceCallback.onPlaybackStopped()
        serviceCallback.onStopNotification()
    }

    private fun handleSkipToNext() {
        if (trackManager.canSkipTrack(ACTION_PLAY_NEXT)) {
            handlePlay()
        } else {
            handleStop()
        }
    }

    private fun handleSkipToPrev() {
        if (playback.getCurrentStreamingPosition() > 3000) {
            playback.onSeekTo(0)
        } else if (trackManager.canSkipTrack(ACTION_PLAY_PREV)) {
            handlePlay()
        }
    }

    private fun publishMetadata(md: MusicModel) {
        val builder = MediaMetadata.Builder()
            .putString(METADATA_TITLE_KEY, md.songName)
            .putString(METADATA_ARTIST_KEY, md.artist)
            .putString(METADATA_ALBUM_KEY, md.album)
            .putLong(METADATA_DURATION_KEY, md.duration)

        serviceCallback.onMetaDataChanged(builder.build())
    }

    override fun onPlaybackCompletion() {
        handleSkipToNext()
    }

    override fun onTrackChangedSeamlessly() {
        if (trackManager.canSkipTrack(ACTION_PLAY_NEXT)) {
            val item = trackManager.getActiveQueueItem() ?: return
            lastKnownPath = item.songPath
            publishMetadata(item)
            serviceCallback.onStartNotification()
        } else {
            handleStop()
        }
    }

    override fun onPlaybackStateChanged(state: Int) {
        val actions = (PlaybackState.ACTION_PLAY
                or PlaybackState.ACTION_PAUSE
                or PlaybackState.ACTION_PLAY_PAUSE
                or PlaybackState.ACTION_SKIP_TO_NEXT
                or PlaybackState.ACTION_SKIP_TO_PREVIOUS
                or PlaybackState.ACTION_STOP
                or PlaybackState.ACTION_SEEK_TO)

        val builder = PlaybackState.Builder()
            .setActions(actions)
            .setState(state, playback.getCurrentStreamingPosition(), 1.0f, SystemClock.elapsedRealtime())

        serviceCallback.onPlaybackStateChanged(builder.build())
    }

    override fun onFocusChanged(resumePlayback: Boolean) {
        if (resumePlayback) {
            handlePlay()
        } else {
            handlePause()
        }
    }
}
